pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -253,43 +253,59 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 		return fmt.Errorf("node %s doesn't have providerID", nodeName)
 	}
 
-	cidrs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
+	podCIDRs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
 	if err != nil {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
-	if len(cidrs) == 0 {
+	if len(podCIDRs) == 0 {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
-	_, cidr, err := net.ParseCIDR(cidrs[0])
-	if err != nil {
-		return fmt.Errorf("failed to parse string '%s' as a CIDR: %v", cidrs[0], err)
+	if _, err := netutils.ParseCIDRs(podCIDRs); err != nil {
+		return err
 	}
-	podCIDR := cidr.String()
 
-	if node.Spec.PodCIDR == podCIDR {
-		klog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDR)
+	skip := false
+
+	if len(podCIDRs) == 1 && node.Spec.PodCIDR == podCIDRs[0] {
```

Comment:
To make the logic easier to understand, you should factor out the "diff" part into a helper function; something like:

// diffNodePodCIDRs returns true if node.Spec contains the
// set of CIDRs assigned to it by the cloud.
func diffNodePodCIDRs(node.Spec, podCIDRs ...) bool {
  ...
}

Reply:
```
if ! diffNodePodCIDRs(...)  {
  // compute new CIDR assignment, update
}
```

Reply:
Added updateNodePodCIDRs to handle this logic.
This function will return true if we need to update, and false if we have the same IPs already present in node.Spec.

Added unit tests for this function.